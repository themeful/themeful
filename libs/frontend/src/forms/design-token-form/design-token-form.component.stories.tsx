import { h } from '@stencil/core'

export default {
  title: 'Forms/Design Token',
  args: {
    first: 'John',
    middle: 'S',
    last: 'Doe',
  },
}

export const designToken = (args) => {
  return (
    <div>
      <div class="header">
        <h1>Design Token Form</h1>
      </div>
      <div class="content">
        <div style={{ position: 'relative', width: '600px' }}>
          <tf-design-token-form {...args} />
        </div>
      </div>
    </div>
  )
}
