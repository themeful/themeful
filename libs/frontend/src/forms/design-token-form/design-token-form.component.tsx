import { Component, h, Prop } from '@stencil/core'

@Component({
  tag: 'tf-design-token-form',
  styleUrl: 'design-token-form.component.scss',
  shadow: true,
})
export class DesignTokenFormComponent {
  /**
   * The first name
   */
  @Prop() first: string

  /**
   * The middle name
   */
  @Prop() middle: string

  /**
   * The last name
   */
  @Prop() last: string

  private getText(): string {
    return `${this.first} ${this.middle} ${this.last}`
  }

  public render(): HTMLTfDesignTokenFormElement {
    return <div>Hello, World! {this.getText()}</div>
  }
}
