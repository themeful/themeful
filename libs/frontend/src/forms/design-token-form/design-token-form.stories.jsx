export default {
  title: 'Forms/Design Token',
  args: {
    first: 'John',
    middle: 'S',
    last: 'Doe',
  },
}

export const designToken = (args) => {
  return <design-token-form {...args}></design-token-form>
}
