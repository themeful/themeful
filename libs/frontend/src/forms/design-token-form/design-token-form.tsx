import { Component, h, Prop } from '@stencil/core'

@Component({
  tag: 'design-token-form',
  styleUrl: 'design-token-form.scss',
  shadow: true,
})
export class DesignTokenForm {
  /**
   * The first name
   */
  @Prop() first: string

  /**
   * The middle name
   */
  @Prop() middle: string

  /**
   * The last name
   */
  @Prop() last: string

  private getText(): string {
    return `${this.first} ${this.middle} ${this.last}`
  }

  render() {
    return <div>Hello, World! I'm {this.getText()}</div>
  }
}
