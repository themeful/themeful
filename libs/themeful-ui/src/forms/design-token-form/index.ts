export * from './design-token-form.component'
