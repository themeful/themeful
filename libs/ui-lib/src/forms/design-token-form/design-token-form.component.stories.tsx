import { propertySelect } from '@properties'
import { h } from '@stencil/core'
import { action } from '@storybook/addon-actions'
import Fragment from 'stencil-fragment'

export default {
  title: 'Forms/Design Token',
  args: {
    samples: [
      {
        formData: {
          groups: ['Base', 'Content', 'More'],
          propertyTypes: propertySelect,
          fields: {
            name: 'Some Token Name',
            group: 'Base',
            type: 'color',
            description: 'Some good description',
          },
        },
        onAction: action('triggered action'),
      },
      {
        formData: {
          identifier: 'dtSomeTokenName',
          groups: ['Base', 'Content', 'More'],
          propertyTypes: propertySelect,
          fields: {
            name: 'Some Token Name',
            group: 'Base',
            type: 'color',
            description: 'Some good description',
          },
        },
        onAction: action('triggered action'),
      },
      {
        formData: {
          groups: ['Base', 'Content', 'More'],
          propertyTypes: propertySelect,
        },
        onAction: action('triggered action'),
      },
    ],
  },
}

export const designToken = ({ samples }) => {
  return (
    <div>
      <div class="header">
        <h1>Design Token Form</h1>
      </div>
      <div class="content content--grid">
        {samples.map((args) => (
          <Fragment>
            <div class="tf-light">
              <tf-design-token-form {...args} />
            </div>
            <div>
              <tf-design-token-form {...args} />
            </div>
          </Fragment>
        ))}
      </div>
    </div>
  )
}
