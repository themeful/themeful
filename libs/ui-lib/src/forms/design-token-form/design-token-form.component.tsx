import { Component, Event, EventEmitter, h, Prop, State } from '@stencil/core'
import {
  DesignTokenFormAction,
  DesignTokenFormData,
  DesignTokenFormFields,
  FormValues,
} from '@typings'
import '../../components/button'
import '../../components/inputs/select-input'
import '../../components/inputs/suggest-input'
import '../../components/inputs/text-input'

@Component({
  tag: 'tf-design-token-form',
  styleUrl: 'design-token-form.component.scss',
  shadow: true,
})
export class DesignTokenFormComponent {
  /** Data for the form */
  @Prop() formData!: DesignTokenFormData

  /** Event emitted when an action is triggered */
  @Event({ composed: false }) action!: EventEmitter<DesignTokenFormAction>

  @State() changed = false
  @State() editMode!: boolean

  private controls: {
    [key: string]: HTMLTfTextInputElement | HTMLTfSelectInputElement | HTMLTfSuggestInputElement
  } = {}

  public componentWillLoad(): void {
    this.editMode = !!this.formData.identifier && true
  }

  public componentDidLoad(): void {
    this.action.emit({ action: 'open' })
  }

  private formValues = (): FormValues =>
    Object.entries(this.controls).reduce((result, [key, control]) => {
      result[key] = control.value
      return result
    }, {})

  private validate = (): Promise<boolean> =>
    Promise.all(Object.values(this.controls).map((control) => control.validate())).then(
      (controls) => controls.every((valid) => valid)
    )

  private dirty = (): Promise<boolean> =>
    Promise.all(Object.values(this.controls).map((control) => control.dirty())).then((controls) =>
      controls.some((valid) => valid)
    )

  private save = async (event: Event): Promise<void> => {
    event.preventDefault()
    Promise.all([this.dirty(), this.validate()]).then(([dirty, valid]) => {
      if (dirty && valid) {
        if (this.editMode) {
          this.action.emit({
            action: 'update',
            identifier: this.formData.identifier,
            fields: this.formValues() as unknown as DesignTokenFormFields,
          })
        } else {
          this.action.emit({
            action: 'create',
            fields: this.formValues() as unknown as DesignTokenFormFields,
          })
        }
      } else if (valid) {
        this.action.emit({ action: 'close' })
      }
    })
  }

  private cancel = (): void => {
    this.action.emit({ action: 'close' })
  }

  private remove = (): void => {
    if (this.editMode) {
      this.action.emit({ action: 'delete', identifier: this.formData.identifier })
    }
  }

  public render(): HTMLTfDesignTokenFormElement {
    return (
      <form class="form" onSubmit={this.save}>
        <h3>{this.editMode ? 'Edit' : 'Create'} Design Token</h3>
        <tf-text-input
          ref={(el: HTMLTfTextInputElement) => (this.controls['name'] = el)}
          label="Name"
          value={this.formData.fields?.name}
          minLength={4}
        />
        <tf-suggest-input
          ref={(el: HTMLTfSuggestInputElement) => (this.controls['group'] = el)}
          label="Group"
          items={this.formData.groups}
          value={this.formData.fields?.group}
          minLength={4}
        />
        <tf-select-input
          ref={(el: HTMLTfSelectInputElement) => (this.controls['type'] = el)}
          label="Type"
          items={this.formData.propertyTypes}
          value={this.formData.fields?.type}
          required
        />
        <tf-text-input
          ref={(el: HTMLTfTextInputElement) => (this.controls['description'] = el)}
          label="Description"
          value={this.formData.fields?.description}
        />
        <div class="form__controls">
          {this.editMode && (
            <tf-button kind="danger" onClick={this.remove}>
              Delete
            </tf-button>
          )}
          <tf-button kind="secondary" onClick={this.cancel}>
            Cancel
          </tf-button>
          <tf-button kind="primary" onClick={this.save} type="submit">
            Save
          </tf-button>
        </div>
      </form>
    )
  }
}
