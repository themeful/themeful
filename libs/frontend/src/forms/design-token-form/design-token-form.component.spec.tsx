import { newSpecPage } from '@stencil/core/testing'
import { DesignTokenFormComponent } from './design-token-form.component'

xdescribe('DesignTokenFormComponent', () => {
  it('renders', async () => {
    const { root } = await newSpecPage({
      components: [DesignTokenFormComponent],
      html: '<tf-design-token-form></tf-design-token-form>',
    })
    expect(root).toEqualHtml(`
      <tf-design-token-form>
        <mock:shadow-root>
          <div>
            Hello, World! I'm
          </div>
        </mock:shadow-root>
      </tf-design-token-form>
    `)
  })

  it('renders with values', async () => {
    const { root } = await newSpecPage({
      components: [DesignTokenFormComponent],
      html: `<tf-design-token-form first="Stencil" last="'Don't call me a framework' JS"></tf-design-token-form>`,
    })
    expect(root).toEqualHtml(`
      <tf-design-token-form first="Stencil" last="'Don't call me a framework' JS">
        <mock:shadow-root>
          <div>
            Hello, World! I'm Stencil 'Don't call me a framework' JS
          </div>
        </mock:shadow-root>
      </tf-design-token-form>
    `)
  })
})
