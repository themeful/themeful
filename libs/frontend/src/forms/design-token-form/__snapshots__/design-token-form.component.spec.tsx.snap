// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DesignTokenFormComponent renders 1`] = `
<tf-design-token-form>
  <mock:shadow-root>
    <form class="form">
      <h3>
        Edit Design Token
      </h3>
      <tf-text-input>
        <mock:shadow-root>
          <label class="text-input">
            <span class="text-input__label">
              Name
            </span>
            <input class="text-input__input" type="text">
            <p class="text-input__hint"></p>
          </label>
        </mock:shadow-root>
      </tf-text-input>
      <tf-suggest-input>
        <mock:shadow-root>
          <label class="suggest-input">
            <span class="suggest-input__label">
              Group
            </span>
            <input autocomplete="off" class="suggest-input__input" list="suggestdatalist1" type="text">
            <datalist id="suggestdatalist1">
              <option value="First"></option>
              <option value="Second"></option>
            </datalist>
            <p class="suggest-input__hint"></p>
          </label>
        </mock:shadow-root>
      </tf-suggest-input>
      <tf-select-input>
        <mock:shadow-root>
          <label class="select-input">
            <span class="select-input__label">
              Type
            </span>
            <select class="select-input__input">
              <option value="first">
                First
              </option>
              <option value="second">
                Second
              </option>
            </select>
            <p class="select-input__hint"></p>
          </label>
        </mock:shadow-root>
      </tf-select-input>
      <tf-text-input>
        <mock:shadow-root>
          <label class="text-input">
            <span class="text-input__label">
              Description
            </span>
            <input class="text-input__input" type="text">
            <p class="text-input__hint"></p>
          </label>
        </mock:shadow-root>
      </tf-text-input>
      <div class="form__controls">
        <tf-button>
          <mock:shadow-root>
            <button class="button danger" type="button">
              <slot></slot>
            </button>
          </mock:shadow-root>
          Delete
        </tf-button>
        <tf-button>
          <mock:shadow-root>
            <button class="button secondary" type="button">
              <slot></slot>
            </button>
          </mock:shadow-root>
          Cancel
        </tf-button>
        <tf-button>
          <mock:shadow-root>
            <button class="button primary" type="submit">
              <slot></slot>
            </button>
          </mock:shadow-root>
          Save
        </tf-button>
      </div>
    </form>
  </mock:shadow-root>
</tf-design-token-form>
`;
